### ISSUE TITLE
RunSettings file not used in TFS VsTest task

### ISSUE THREAD

**Issue by shortstuffsushi** [Microsoft/vsts-tasks#8065]
## Environment
TFS On Prem, Version 16.131.27701.1
Agent - Windows Server 2012R2, agent version 2.131.0

## Issue Description

While running the vstest task, the `.runsettings` file I've provided is being ignored and replaced by an autogenerated version, causing the coverage report to include assemblies I don't want. These are what I believe to be the relevant settings, let me know if you need anything more.

![image](https://user-images.githubusercontent.com/1016147/44278532-d5cf2180-a213-11e8-91d1-31f0c175c740.png)

### Error logs

These are what I believe to be the relevant logs, let me know if you need anything more.

```
2018-08-17T15:49:03.0063756Z ==============================================================================
2018-08-17T15:49:03.8104045Z Run the tests locally using vstest.console.exe
2018-08-17T15:49:03.8104260Z ========================================================
2018-08-17T15:49:03.8104856Z Test selector : Test assemblies
2018-08-17T15:49:03.8105068Z Test assemblies : **\*Tests.dll,!**\*TestAdapter.dll,!**\obj\**
2018-08-17T15:49:03.8105262Z Test filter criteria : null
2018-08-17T15:49:03.8105432Z Search folder : C:\agent\_work\181\s
2018-08-17T15:49:03.8105609Z Run settings file : C:\agent\_work\181\s\CodeCoverage.runsettings
^^^^^^ Using my settings file
2018-08-17T15:49:03.8105859Z Run in parallel : false
2018-08-17T15:49:03.8106037Z Run in isolation : false
2018-08-17T15:49:03.8106977Z Path to custom adapters : null
2018-08-17T15:49:03.8107159Z Other console options : null
2018-08-17T15:49:03.8107319Z Code coverage enabled : true
2018-08-17T15:49:03.8107664Z Rerun failed tests: false
2018-08-17T15:49:03.9133771Z VisualStudio version selected for test execution : toolsInstaller
2018-08-17T15:49:04.1860621Z ========================================================
2018-08-17T15:49:04.4973145Z [command]C:\agent\_work\_tool\VsTest\15.9.0-preview-20180807-05\x64\tools\net451\Common7\IDE\Extensions\TestPlatform\vstest.console.exe @C:\Windows\SERVIC~2\NETWOR~1\AppData\Local\Temp\11251751-a235-11e8-8719-85a908d74f4f.txt
2018-08-17T15:49:04.7074312Z Microsoft (R) Test Execution Command Line Tool Version 15.9.0-preview-20180807-05
2018-08-17T15:49:04.7074631Z Copyright (c) Microsoft Corporation.  All rights reserved.
2018-08-17T15:49:04.7074717Z 
2018-08-17T15:49:04.7074811Z vstest.console.exe 
2018-08-17T15:49:04.7074939Z "C:\agent\_work\181\s\MyProjectName\bin\Release\MyProjectName.Tests.dll"
2018-08-17T15:49:04.7075151Z /Settings:"C:\Windows\SERVIC~2\NETWOR~1\AppData\Local\Temp\11441100-a235-11e8-8719-85a908d74f4f.runsettings"
^^^^^^ Not using my settings file
2018-08-17T15:49:04.7075274Z /logger:"trx"
2018-08-17T15:49:04.7075379Z /TestAdapterPath:"C:\agent\_work\181\s"
2018-08-17T15:49:04.8381749Z Starting test execution, please wait...
```

Initially while reviewing those logs, I assumed that the temp file it was creating may just be a copy of my coverage file, which is as follows:
```
<?xml version="1.0" encoding="utf-8"?>
<RunSettings>
  <DataCollectionRunSettings>
    <DataCollectors>
      <DataCollector friendlyName="Code Coverage" uri="datacollector://Microsoft/CodeCoverage/2.0" assemblyQualifiedName="Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector, Microsoft.VisualStudio.TraceCollector, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
        <Configuration>
          <CodeCoverage>
            <ModulePaths>
              <Include>
                <ModulePath>.*\\MyProjectName.*\.dll$</ModulePath>
              </Include>
              <Exclude>
                <ModulePath>.*Tests.dll$</ModulePath>
              </Exclude>
            </ModulePaths>
          </CodeCoverage>
        </Configuration>
      </DataCollector>
    </DataCollectors>
  </DataCollectionRunSettings>
</RunSettings>
```

However, I was able to catch one of the generated `.runsetting` files it's creating, which is effectively empty:

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<RunSettings>
  <DataCollectionRunSettings>
    <DataCollectors>
      <DataCollector friendlyName="Code Coverage" uri="datacollector://Microsoft/CodeCoverage/2.0" assemblyQualifiedName="Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector, Microsoft.VisualStudio.TraceCollector, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
        <Configuration>
          <CodeCoverage>
            <UseVerifiableInstrumentation>False</UseVerifiableInstrumentation>
          </CodeCoverage>
        </Configuration>
      </DataCollector>
    </DataCollectors>
  </DataCollectionRunSettings>
</RunSettings>
```

As a final side note (and this may make sense to make as a separate issue) the test cleanup deletes the `.runsetting` file it creates in the temp directory, but leaves a `.txt` file containing the path to the test dll, settings file location, logger, and test adapter path. It seems likely this is intended to be cleaned up as well, but was missed.

**Comment by acesiddhu** [comment]
@shortstuffsushi I see that you have also checked Code Coverage Enabled checkbox. Can you un check and try again. Saying this because you already have enabled Code coverage in your runsettings and you don't need to use this task option.
Regarding cleanup, with recent changes in our task (that will flow in vsts) we are cleaning up all generated files. so this should be fixed.

**Comment by shortstuffsushi** [comment]
Hey @acesiddhu, I had thought you may ask that, but chose to not include those results because they seem like probably an unrelated thing. Unfortunately, running *without* that checked is even worse, as it results in all tests failing with the following:

```
Failed   MyTestMethod
Error Message:
 Test method MyTestMethod threw exception: 
System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.VisualStudio.CodeCoverage.Shim, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
Stack Trace:
   at MyClass..ctor(...)
   at MyTestMethod() in C:\...\MyTestFIle.cs:line 40
```

**Comment by acesiddhu** [comment]
@shortstuffsushi this was surely a bug on our side which we have fixed @prawalagarwal can you confirm. in the mean time you can change your run settings to include in your code coverage configuration node.

<Configuration>
          <CodeCoverage>
            <UseVerifiableInstrumentation>False</UseVerifiableInstrumentation>
          </CodeCoverage>
        </Configuration>

**Comment by shortstuffsushi** [comment]
I'm not sure what you're asking me to include. Do you want me to put the text `false` inside one of the nodes? Could you give a more complete example of what you're asking for?

**Comment by acesiddhu** [comment]
Sorry GH ate my comment because it an xml
here is the runsetting u can use

```
<?xml version="1.0" encoding="utf-8"?>
<RunSettings>
  <DataCollectionRunSettings>
    <DataCollectors>
      <DataCollector friendlyName="Code Coverage" uri="datacollector://Microsoft/CodeCoverage/2.0" assemblyQualifiedName="Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector, Microsoft.VisualStudio.TraceCollector, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
        <Configuration>
          <CodeCoverage>
           <UseVerifiableInstrumentation>False</UseVerifiableInstrumentation>
            <ModulePaths>
              <Include>
                <ModulePath>.*\\MyProjectName.*\.dll$</ModulePath>
              </Include>
              <Exclude>
                <ModulePath>.*Tests.dll$</ModulePath>
              </Exclude>
            </ModulePaths>
          </CodeCoverage>
        </Configuration>
      </DataCollector>
    </DataCollectors>
  </DataCollectionRunSettings>
</RunSettings>
```

**Comment by shortstuffsushi** [comment]
Ok, adding the `UseVerifiableInstrumentation = False` seems to have worked, could you explain why?

**Comment by acesiddhu** [comment]
you are using visual studio test tools installer task in your definition. in case that task is used useverifiable property needs to be set to false because in xcopy mode we don't GAC the above dll 'Microsoft.VisualStudio.CodeCoverage.Shim (this is a dependency which needs to be loaded when useverifiable is set to true)

marking this property false ensures it doesn't use this particular dll.

**Comment by acesiddhu** [comment]
Hi @shortstuffsushi I am closing the issue. Thanks

**Comment by shortstuffsushi** [comment]
@acesiddhu I count three open issues here.

1. Use Code Coverage Enabled ignores user `.runsettings` files
2. `.txt` generated files aren't being cleaned up
3. Disabling Code Coverage Enabled causes errors because `Microsoft.VisualStudio.CodeCoverage.Shim` is unavailable

I'm at a point where I can work around these, but ideally they'd be fixed right? Are you tracking them internally or something? It seems like this issue isn't resolved.

**Comment by acesiddhu** [comment]
Hi @shortstuffsushi sorry for not replying to this before. I tested all the scenario with our latest code and here are the findings

1) This seems to be fixed in latest code base. if I provide .runsettings files with Code coverage on, I see it is honoring the runsettings I provide

2) The txt files are getting cleaned up.

3) this is also fixed in latest code base.

Let me know if you any other issues

**Comment by shortstuffsushi** [comment]
Is there a way for me to check this out on my end / on our local TFS instance? Will the updates you mentioned be part of an upcoming release?

**Comment by acesiddhu** [comment]
there is no way to test the changes in ur local tfs because vstest task might be incompatible with your current tfs version. the fixes will be in upcoming Azuredevops server release.
